### PR TITLE
Iox #285 added unit test for roudi memory block manager provider

### LINKED ISSUE #454: Enhancing Tests in Roudi
## Brief feature description
Introduce Module tests for RouDi component.

## Detailed information
Increase the code coverage.

Classes in RouDi components which needs test cases to be added.
In file roudi.cpp, class 
- [ ] RouDi.

In file roudi_process.cpp class 
- [x] RouDiProcess
- [x] PortManager
- [ ] ProcessManager
- [ ] RoudiMemoryManager

In file roudi_lock.cpp class 
- [ ] RouDiLock

### PR DESCRIPTION
**Pull request by nihalchari**
## Pre-Review Checklist for the PR Author

1. [ ] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [ ] Commits messages are according to this [guideline][commit-guidelines]
    - [ ] Commit messages have the issue ID (`iox-#123 commit text`)
    - [ ] Commit messages are signed (`git commit -s`)
    - [ ] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [ ] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [ ] Relevant issues are linked
1. [ ] Add sensible notes for the reviewer
1. [ ] All checks have passed
1. [ ] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Notes for Reviewer

code contributed by @JimmyBch
created draft PR as contributor has open points to discuss (commented code)

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [ ] All checkboxes in the PR checklist are checked or crossed out
1. [ ] Merge

## References

- Partly closes #454

### CODE REVIEW

**Review by dkroenke — COMMENTED** [review]
Hi @JimmyBch and welcome to iceoryx! 
Thanks for contributing, i see that you have commented out the mock functions for the memory provider. I guess this is the discussion point here right?
If you have question regarding the contribution process just let us know.

**Comment by JimmyBch** [comment]
Hi @dkroenke, yes indeed the code let in comment is for discussion.

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp [hunk 1]
```diff
@@ -71,8 +71,21 @@ TEST_F(RouDiMemoryManager_Test, CallingCreateMemoryWithMemoryProviderSucceeds)
...
 
-    EXPECT_CALL(memoryBlock1, destroyMock());
-    EXPECT_CALL(memoryBlock2, destroyMock());
+//    EXPECT_CALL(memoryBlock1, destroyMock());
+//    EXPECT_CALL(memoryBlock2, destroyMock());
+}
+
+TEST_F(RouDiMemoryManager_Test, CallingCreateMemoryWithMemoryProviderError)
+{
+    sut.addMemoryProvider(&memoryProvider1);
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:80]: Why does this lead to an error? I am not fully in the loop functionality-wise, a short explanation would be nice.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 2]
```diff
@@ -12,7 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//#define protected public
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:15]: If commented out it has to be removed. It is really dangerous to use since changing the access specifier may change the data layout leading to undefined bevahiour during data access.

Do not introduce those kind of defines in new (test) code.

Legacy tests where this is still done need to be reworked in another issue.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 3]
```diff
@@ -12,7 +12,9 @@
...
 #include "iceoryx_posh/roudi/memory/memory_provider.hpp"
+//#undef protected
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:17]: remove line

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 4]
```diff
@@ -38,10 +40,17 @@ class MemoryProviderFailingCreation : public iox::roudi::MemoryProvider
     {
         return iox::cxx::error<MemoryProviderError>(MemoryProviderError::MEMORY_DESTRUCTION_FAILED);
     }
+
+    static const char* getErrorStringMemoryProvider(const MemoryProviderError error)
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:44]: Considering the redirection internally, is this function really useful? Looks like it could be removed and replaced by the call directly.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 5]
```diff
@@ -64,14 +73,49 @@ class MemoryProvider_Test : public Test
...
+    //    EXPECT_CALL(sut, createMemoryMock(COMMON_SETUP_MEMORY_SIZE, COMMON_SETUP_MEMORY_ALIGNMENT)).Times(1);
 
-        EXPECT_CALL(sut, destroyMemoryMock());
+    //    EXPECT_CALL(sut, destroyMemoryMock());
         EXPECT_CALL(memoryBlock1, destroyMock());
 
         return sut.create();
     }
 
+    static const int16_t nTestCase = 13;
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:84]: ``int16_t`` is ok, but it really buys us nothing here, could be ``int32_t`` or ``uint32_t`` as well. Note that smaller types can even be detrimental wrt. execution time when the native machine word length is larger (due to the way the hardware has to access those bytes).

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 6]
```diff
@@ -302,3 +344,13 @@ TEST_F(MemoryProvider_Test, SegmentIdValueAfterDestructionIsUnset)
 
     EXPECT_THAT(sut.segmentId().has_value(), Eq(false));
 }
+
+TEST_F(MemoryProvider_Test, GetErrorString)
+{
+
+    for(int16_t i = 0; i < nTestCase;i++)
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:352]: No. Not if ``nTestCase`` is signed (which it is). And I would not introduce unsigned types in tests where it is not necessary. This has the nice side effect of sparing us from the Us everywhere without breaking (useless) conventions. ;-)

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 7]
```diff
@@ -302,3 +346,12 @@ TEST_F(MemoryProvider_Test, SegmentIdValueAfterDestructionIsUnset)
 
     EXPECT_THAT(sut.segmentId().has_value(), Eq(false));
 }
+
+TEST_F(MemoryProvider_Test, GetErrorString)
+{
+    for(int16_t i = 0; i < nTestCase;i++)
+    {
+        const char * result = MemoryProviderFailingCreation::getErrorStringMemoryProvider(m_testCombinationMemoryProviderError[i]);
```
Comment (MatthiasKillat) [inline iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:354]: Can we not just call ``MemoryProvider::getErrorString`` directly here? For me it seems the only thing the test does is check the translation of enum values to strings. This is ok, but if those enum value or string array is not used elsewhere (did not fully check), they may be better placed in the test itself and not in the test fixture.

I think this test is not very meaningful in itself. What is needed to test failed memory creation and is it done in a sufficient manner elsewhere?

**Review by MatthiasKillat — CHANGES_REQUESTED** [review]
Commented out code should be removed or explained why it is there. I am not sure the tests were improved by those changes, can you elaborate what the benefit of these changes is?

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JimmyBch) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp]: I removed that part :)

File: iceoryx_posh/test/mocks/roudi_memory_provider_mock.hpp [hunk 8]
```diff
@@ -35,29 +35,29 @@ class MemoryProviderTestImpl : public iox::roudi::MemoryProvider
     iox::cxx::expected<void*, iox::roudi::MemoryProviderError> createMemory(const uint64_t size,
                                                                             const uint64_t alignment) noexcept override
     {
-        if (m_mockCallsEnabled)
+        /*if (m_mockCallsEnabled)
```
[the remark this answers is not in the document]
Comment (JimmyBch) [reply iceoryx_posh/test/mocks/roudi_memory_provider_mock.hpp:38]: I wanted to discuss the code whether it was really necessary or not. I can also remove it as well.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp (hunk 7 above)
Comment (JimmyBch) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:352]: Yes, indeed it is only testing the translation. But in order to improve the coverage, that's the only I found. We cannot just call the method because it is protected. That's why I implemented another method that's public and can be called in test.

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp (hunk 1 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:99]: have a look at the line below ;)

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp [hunk 9]
```diff
@@ -70,9 +92,17 @@ TEST_F(RouDiMemoryManager_Test, CallingCreateMemoryWithMemoryProviderSucceeds)
     sut.addMemoryProvider(&memoryProvider2);
 
     EXPECT_THAT(sut.createAndAnnounceMemory().has_error(), Eq(false));
+}
 
-    EXPECT_CALL(memoryBlock1, destroyMock());
-    EXPECT_CALL(memoryBlock2, destroyMock());
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:75]: I would not remove this. How do you check that the memory destruction was triggered?

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp [hunk 10]
```diff
@@ -89,9 +119,7 @@ TEST_F(RouDiMemoryManager_Test, RouDiMemoryManagerDTorTriggersMemoryProviderDest
         RouDiMemoryManager sutDestroy;
         sutDestroy.addMemoryProvider(&memoryProvider1);
         sutDestroy.createAndAnnounceMemory();
-        EXPECT_CALL(memoryBlock1, destroyMock()).Times(1);
     }
-    EXPECT_CALL(memoryBlock1, destroyMock()).Times(0);
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:94]: same here. It's important to check that the destruction was triggered

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 11]
```diff
@@ -117,7 +157,6 @@ TEST_F(MemoryProvider_Test, AddMemoryBlockExceedsCapacity)
 
 TEST_F(MemoryProvider_Test, CreateWithoutMemoryBlock)
 {
-    EXPECT_CALL(sut, createMemoryMock(_, _)).Times(0);
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:120]: I wouldn't remove this. We need to check if our invariants hold and that is one of these.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp [hunk 12]
```diff
@@ -175,19 +213,13 @@ TEST_F(MemoryProvider_Test, CreateAndAnnounceWithMultipleMemoryBlocks)
     EXPECT_CALL(memoryBlock1, alignmentMock()).WillRepeatedly(Return(MEMORY_ALIGNMENT_1));
     EXPECT_CALL(memoryBlock2, sizeMock()).WillRepeatedly(Return(MEMORY_SIZE_2));
     EXPECT_CALL(memoryBlock2, alignmentMock()).WillRepeatedly(Return(MEMORY_ALIGNMENT_2));
-    EXPECT_CALL(sut, createMemoryMock(MEMORY_SIZE_1 + MEMORY_SIZE_2, std::max(MEMORY_ALIGNMENT_1, MEMORY_ALIGNMENT_2)))
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:178]: same here. Why did you remove this checks?

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp (hunk 7 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:341]: you can just add `using MemoryProvider::getErrorString;` in the public section of `MemoryProviderFailingCreation` to access it.

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp (hunk 10 above)
Comment (JimmyBch) [reply iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:94]: Thanks for the comment, i was wondering why was it useful ? :)

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp (hunk 7 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:341]: I see. It is for coverage purposes. Now I understand the inheritance step, so feel free to ignore my comment regarding the necessity of this function.

This is somewhat unfortunate, because we are testing a protected function (i.e. implementation detail) instead of forcing the error by using the public API (this might be not easy in this case, depending on the error).

Ideally the error should be created by some use of public API functions, which in turn leads to the coverage as well. I do not know whether this is possible here.

The problem is, that we are increasing the coverage without actually testing whether the errors are actually under the conditions we expect. This gives a false sense of security/coverage.

@elBoberido We need to evaluate how it can be usefully tested (maybe not in this PR, I am not sure).

File: iceoryx_posh/test/mocks/roudi_memory_provider_mock.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_posh/test/mocks/roudi_memory_provider_mock.hpp]: I have little knowledge about those tests, maybe @elBoberido knows the details. It should not be merged with these commented sections in my opinion. The question is: if they are removed, are the tests still useful (not only from a coverage point of view)?

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp (hunk 10 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:94]: The MemoryProvider tells the MemoryBlock that the memory will be destroyed, which means they must call the destructor of the object they store. These functions are pure virtual and needs to be implemented by the concrete memory block. We just need to check that the MemoryProvider works correctly and the chain of destruction is not broken. Unfortunately, we do not yet have enough documentation for this. I plan to add some when the SegmentManager is refactored.

File: iceoryx_posh/test/mocks/roudi_memory_provider_mock.hpp [hunk 13]
```diff
@@ -35,29 +35,29 @@ class MemoryProviderTestImpl : public iox::roudi::MemoryProvider
...
         {
             createMemoryMock(size, alignment);
-        }
+        }*/
 
         dummyMemory = static_cast<uint8_t*>(iox::cxx::alignedAlloc(alignment, size));
         return iox::cxx::success<void*>(dummyMemory);
     }
-    MOCK_METHOD2(createMemoryMock, void(uint64_t, uint64_t));
+    //MOCK_METHOD2(createMemoryMock, void(uint64_t, uint64_t));
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_posh/test/mocks/roudi_memory_provider_mock.hpp]: I think we can keep the mock methods, can we not? We may or may not check for calls, but is it a problem to have them?

Possibly due to unexpected mock calls? But if a expecting a call is important for the test, we need those. And if not, it should just be removed as an unnecessary implementation detail which is of no concern for the test.

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp (hunk 1 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:99]: I did, before the comment (I think you mean the ASSERT). It is still not clear to me why adding a ``memoryProvider1`` must lead to an error and neither the test name nor the error code are really providing this information. Maybe I have to take a look at ``memoryProvider1``.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp (hunk 4 above)
Comment (MatthiasKillat) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp:43]: OK. As commented elsewhere, it is due to protected access.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp (hunk 7 above)
Comment (JimmyBch) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp]: Hi Matthias, we are testing all the input and checking whether the outputs are correct. So, on a level of unit test, I would say it is sufficient.

**Comment by JimmyBch** [comment]
@MatthiasKillat @elBoberido @dkroenke Hi guys, do you think something else should be changed ? If not, can we create an official PR in order to merge ?

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp [hunk 14]
```diff
@@ -109,6 +142,15 @@ TEST_F(RouDiMemoryManager_Test, AddMemoryProviderExceedsCapacity)
...
-/// @todo Add TEST_F(RouDiMemoryManager_Test, StreamOperatorTranslationIsCorrect)
+TEST_F(RouDiMemoryManager_Test, OperatorTest)
+{
+    for(int16_t i = 0; i < nbTestCase; i++)
+    {
+        iox::log::LogStream logStream(loggerMock);
+        iox::log::LogStream logStreamResult(loggerMockResult);
+        logStream << m_testCombinationRoudiMemoryManagerError[i];
+        logStreamResult << m_testResultOperatorMethod[i];
+        EXPECT_EQ(*::testing::PrintToString(logStream).c_str(), *::testing::PrintToString(logStreamResult).c_str());
```
Comment (elBoberido) [inline iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:153]: this is overly complicated. Theoretically you should be able to do a `EXPECT_THAT(loggerMock.m_logs[0].message, Eq(m_testResultOperatorMethod[i]));`

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp (hunk 1 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:102]: It doesn't make sense to have a memory provider without a memory block because the provider calculates the required memory from the contained blocks.
If no memory block were added to the memory provider it's usually an error, because the developer forgot to add one.

File: iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp (hunk 7 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_memory_provider.cpp]: @MatthiasKillat in this special case it's just a conversion method to get the string representation of the enum. It might even be better to implement the ostream operator for `MemoryProviderError`.

So this test doesn't try to trigger any error, it just checks if the string translation works.

**Comment by elBoberido** [comment]
@JimmyBch yes, of course. Just create a PR

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp (hunk 14 above)
Comment (JimmyBch) [reply iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:153]: Theoretically, but I am getting a segmentation fault.

**Comment by dkroenke** [comment]
> @JimmyBch yes, of course. Just create a PR

I'm not sure if Jimmy can convert this to a PR maybe the committers or the creator of the PR @nihalchari can do this.

**Comment by nihalchari** [comment]
> 
> 
> > @JimmyBch yes, of course. Just create a PR
> 
> I'm not sure if Jimmy can convert this to a PR maybe the committers or the creator of the PR @nihalchari can do this.

converted to PR

**Comment by nihalchari** [comment]
By mistake got closed, reopned again

**Comment by JimmyBch** [comment]
@elBoberido @dkroenke @MatthiasKillat Do you mind to give your approvals for this PR in order to merge it ? :)

**Review by dkroenke — DISMISSED** [review]

**Comment by dkroenke** [comment]
@nihalchari please add the PR Template to your Pull-Request description. Did this not come up when you have created the PR?

File: iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp (hunk 14 above)
Comment (elBoberido) [reply iceoryx_posh/test/moduletests/test_roudi_memory_manager.cpp:153]: The reason for this is that the LogStream object sends the data in its destructor to the logger. Since you don't use a temporary object, the logger receives the data only at the end of the scope, which is too late.

You have three options
- use `logStram.Flush()` to send the data to the logger
- move the following to an own scope
```
iox::log::LogStream logStream(loggerMock);
logStream << m_testCombinationRoudiMemoryManagerError[i];
```
- use a temporary object `iox::log::LogStream(loggerMock) << "" << m_testCombinationRoudiMemoryManagerError[i];` (unfortunately there is a flaw in the Logger and the first item to log with a temporary object must be a type with built in logging support like an empty string`""`)

Additionally, the logger mock keeps all messages until destruction, so you either have to move the logger mock into the for loop or have to check for `loggerMock.m_logs[i].message`. There should also be a sanity check that `loggerMock.m_logs` is large enough to index.

I hope I could give you a better understanding how the logging stuff works

**Comment by elBoberido** [comment]
@JimmyBch could you have a final look at the LogSteam test? This tests get easy copy and pasted so it would be nice to have this changed. Especially because `*::testing::PrintToString(logStream).c_str()` doesn't print the actual string, but just the lucky number 7 ;)

**Comment by JimmyBch** [comment]
@elBoberido I updated the logstream test. I used the first option with .flush. Hope, it is ok for you :)

**Review by elBoberido — COMMENTED** [review]
we are getting closer :D

You need to update your commit messages. Some of them don't have the correct format. This reminds me that we need some pre-commit hooks to check for this. Especially for new contributors it's something easily done wrong.

Since you need to push anyway, could you add an `ASSERT_THAT(loggerMock.m_logs.size(), Eq(i + 1U));` to prevent a segfault if something goes wrong with the logger? My inner chicken would be grateful.

**Comment by elBoberido** [comment]
@JimmyBch could you also pleas go through the checklist of the PR author?

**Comment by JimmyBch** [comment]
@elBoberido Commit messages changed, ASSERT_THAT added :)

**Review by elBoberido — APPROVED** [review]

**Review by dkroenke — DISMISSED** [review]

**Comment by elBoberido** [comment]
@JimmyBch we are getting closer, but we are not quite there yet. I just noticed that you used the PR number instead of the issue number for your commit messages. They must start with `iox-#285` like the name of the branch

**Review by elBoberido — CHANGES_REQUESTED** [review]
update commit messages

**Comment by dkroenke** [comment]
> @JimmyBch we are getting closer, but we are not quite there yet. I just noticed that you used the PR number instead of the issue number for your commit messages. They must start with `iox-#285` like the name of the branch

No that's ok that the commit messages links to https://github.com/eclipse-iceoryx/iceoryx/issues/454 because https://github.com/eclipse-iceoryx/iceoryx/issues/285 is not related to this PR.

**Comment by elBoberido** [comment]
@dkroenke I just noticed that #454 would also fit, but it doesn't match the branch name and there is also one commit with #285. What to use here?

**Comment by dkroenke** [comment]
> @dkroenke I just noticed that #454 would also fit, but it doesn't match the branch name and there is also one commit with #285. What to use here?

https://github.com/eclipse-iceoryx/iceoryx/pull/478/commits/edeb0a6b17bf9c17f8f8e8329ca48f01af3647de is wrongly referenced and needs to be linked to #454

**Comment by elBoberido** [comment]
> @JimmyBch [edeb0a6](https://github.com/eclipse-iceoryx/iceoryx/commit/edeb0a6b17bf9c17f8f8e8329ca48f01af3647de) is wrongly referenced and needs to be linked to #454

@JimmyBch since there is a merge in the commit history, it might be easier to rebase to master and force commit to change that one commit message

**Comment by JimmyBch** [comment]
Is it really an issue to have it link to #285 as the issue as well mention to add unit test ?

**Comment by JimmyBch** [comment]
@elBoberido @dkroenke I couldn't succeed to change the commit message due to the merge commit. Is it a big issue ?

**Review by dkroenke — APPROVED** [review]
I guess the branch name can be as is. Commits are now ok.
@nihalchari Would you please go through the PR checklist?

**Comment by elBoberido** [comment]
@JimmyBch did you try to rebase? Not that big issue, but if it could easily be solved by a rebase, it would be nice.

**Comment by JimmyBch** [comment]
@elBoberido I tried with a rebase. But, I have not been able to do it.

**Comment by elBoberido** [comment]
> @elBoberido I tried with a rebase. But, I have not been able to do it.

@JimmyBch did you try `git rebase master`? Merge conflicts can be fixed with `git mergetool` followed with a `git rebase --continue` and maybe the merge commit needs to be skipped with `git rebase --skip` because it might be empty

**Comment by JimmyBch** [comment]
> > @elBoberido I tried with a rebase. But, I have not been able to do it.
> 
> @JimmyBch did you try `git rebase master`? Merge conflicts can be fixed with `git mergetool` followed with a `git rebase --continue` and maybe the merge commit needs to be skipped with `git rebase --skip` because it might be empty

I will give it another try.

**Comment by JimmyBch** [comment]
> > > @elBoberido I tried with a rebase. But, I have not been able to do it.
> > 
> > 
> > @JimmyBch did you try `git rebase master`? Merge conflicts can be fixed with `git mergetool` followed with a `git rebase --continue` and maybe the merge commit needs to be skipped with `git rebase --skip` because it might be empty
> 
> I will give it another try.

Didn't succeed again. From my understanding, normally I should do a git rebase -i HEAD~15. But, I have not been able to change the message name.

**Comment by elBoberido** [comment]
> Didn't succeed again. From my understanding, normally I should do a git rebase -i HEAD~15. But, I have not been able to change the message name.

No at first you have to rebase to master, e.g.
```
git checkout master
git pull master
git checkout your-branch
git rebase master
```
then the stuff with `git mergetool`, etc. I mentioned above if there are merge conflicts. Once they are fixed you can do a `git rebase -i HEAD~14`

**Comment by JimmyBch** [comment]
> > Didn't succeed again. From my understanding, normally I should do a git rebase -i HEAD~15. But, I have not been able to change the message name.
> 
> No at first you have to rebase to master, e.g.
> 
> ```
> git checkout master
> git pull master
> git checkout your-branch
> git rebase master
> ```
> 
> then the stuff with `git mergetool`, etc. I mentioned above if there are merge conflicts. Once they are fixed you can do a `git rebase -i HEAD~14`

Hi,

I cannot do it again. I am not able to find my commit to change it.

**Comment by mossmaurice** [comment]
> Update: I found my commit but when I try to change the message, it is already with the correct issue number :(

If everything false, you can also try to start with a new branch an manually add your changes with a difftool like [meld](https://meldmerge.org/).

I can recommend this [git game](https://learngitbranching.js.org/), that's were got my git knowledge from ;-)

@marthtz has a lot of git experience and might be able to help you.

**Comment by elBoberido** [comment]
@JimmyBch from the commit history it seems you merged the remote branch after doing the rebase instead of doing a force push to overwrite the remote branch. A rebase to master should remove all the merge commits.

**Comment by elfenpiff** [comment]
@nihalchari Could you please check the whole pre review checklist right after you created the pull request. In this pull request are multiple commits not signed off, please sign them off!

**Review by elfenpiff — CHANGES_REQUESTED** [review]
Please sign off all commits!

**Comment by JimmyBch** [comment]
> @nihalchari Could you please check the whole pre review checklist right after you created the pull request. In this pull request are multiple commits not signed off, please sign them off!

Strange because I am always adding the signed option in my commit.

**Comment by JimmyBch** [comment]
@elBoberido @dkroenke Due to my struggle to change the commit message, this PR will be declined and another one will be created with correct commit messages.

**Comment by elBoberido** [comment]
@JimmyBch a rebase to master should enable you to change all commits. Did you try that? If yes, what were the problems you encountered?

**Comment by nihalchari** [comment]
new PR is created https://github.com/eclipse-iceoryx/iceoryx/pull/487, closing this
